File: sdks/java/endurance-tests/src/test/java/inc/reactor/sdk/endurance/VideoPublishSteadyTest.java
package inc.reactor.sdk.endurance;

import inc.reactor.sdk.Reactor;
import inc.reactor.sdk.Track;
import java.time.Duration;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

/**
 * Publish once and hold it, streaming for the whole run.
 *
 * <p>The opposite shape from every churn scenario. A leak tied to <em>elapsed streaming time or
 * frame count</em> rather than to churn count would not necessarily show up in a churn scenario
 * even run forever, which makes this a distinct category rather than a variant.
 */
final class VideoPublishSteadyTest {

    private static final int WIDTH = 320;
    private static final int HEIGHT = 240;

    @Test
    @DisplayName("video publish steady")
    void videoPublishSteady() throws Exception {
        String apiKey = Endurance.apiKey();
        Endurance.run(
                "video-publish-steady",
                "publish once and hold it, pushing video for the whole run — no pause, no unpublish, "
                        + "no reconnect.",
                false,
                run -> {
                    Reactor reactor = Reactor.open(Endurance.options(apiKey));
                    try {
                        reactor.connect().join();
                        Track input = reactor.track("webcam");
                        input.publish().join();
                        byte[] frame = new byte[WIDTH * HEIGHT * 4];

                        while (run.keepGoing()) {
                            // A cycle is a second of video rather than one frame: at 30fps a per-frame sample
                            // would be more measuring than streaming.
                            for (int index = 0; index < 30 && run.keepGoing(); index++) {
                                input.pushFrame(frame, WIDTH, HEIGHT);
                                Endurance.pause(Duration.ofMillis(33));
                            }

                            // This scenario's own invariant: the publish must survive the whole run, and a
                            // session that dropped out from under it would make every later reading meaningless.
                            if (!input.isPublished()) {
                                throw new AssertionError("the publish did not survive the run");
                            }
                            run.endOfCycle();
                        }
                        input.unpublish();
                    } finally {
                        try {
                            reactor.disconnect().join();
                        } catch (RuntimeException alreadyGone) {
                            // The report still has to be written.
                        }
                        reactor.close();
                    }
                });
    }
}

File: sdks/android/endurance-tests/src/androidTest/kotlin/inc/reactor/sdk/android/endurance/VideoPublishSteadyTest.kt
package inc.reactor.sdk.android.endurance

import androidx.test.ext.junit.runners.AndroidJUnit4
import kotlinx.coroutines.delay
import kotlinx.coroutines.runBlocking
import org.junit.Test
import org.junit.runner.RunWith
import java.nio.ByteBuffer

/**
 * Publish once and hold it, streaming for the whole run.
 *
 * The opposite shape from every churn scenario. A leak tied to *elapsed streaming time or frame
 * count* rather than to churn count would not necessarily show up in a churn scenario even run
 * forever, which makes this a distinct category rather than a variant.
 */
@RunWith(AndroidJUnit4::class)
internal class VideoPublishSteadyTest {
    private companion object {
        const val WIDTH = 320
        const val HEIGHT = 240

        /** Roughly 30fps. Fast enough to be streaming, slow enough not to be a benchmark. */
        const val FRAME_INTERVAL_MS = 33L
    }

    @Test
    fun videoPublishSteady() =
        runBlocking {
            val apiKey = Endurance.apiKey()
            Endurance.run(
                "video-publish-steady",
                "publish once and hold it, pushing video for the whole run — no pause, no " +
                    "unpublish, no reconnect.",
                liveClientsBaselineZero = false,
            ) { run ->
                val reactor = Endurance.connected(apiKey)
                try {
                    val input = reactor.track("webcam")
                    input.publish()
                    // One buffer, reused. Allocating per frame would make the allocator the thing
                    // under test rather than the binding.
                    val frame = ByteBuffer.allocateDirect(WIDTH * HEIGHT * 4)

                    while (run.keepGoing()) {
                        frame.rewind()
                        input.pushFrame(frame, WIDTH, HEIGHT)
                        delay(FRAME_INTERVAL_MS)

                        // This scenario's own invariant: it publishes once, so a slot that stopped
                        // being published means something tore it down behind the caller's back.
                        if (!input.published) {
                            throw AssertionError("the held track stopped being published mid-run")
                        }
                        run.endOfCycle()
                    }
                } finally {
                    runCatching { reactor.disconnect() }
                    reactor.close()
                }
            }
        }
}
